src/V3SplitVar.cpp

```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+void SplitVarVisitor::visit(AstVar* nodep) {
+    m_lastVar = nodep;
+}
+
+void SplitVarVisitor::visit(AstPragma* pragmap) {
+    if (pragmap->pragType() != AstPragmaType::SPLIT_VAR) return;  // nothing to do
+    if (!m_lastVar) {
+        pragmap->v3warn(SPLITVAR, "Stray pragma of split_var is detected.");
+    } else if (!canSplit(m_lastVar)) {
+        pragmap->v3warn(SPLITVAR,
+                        "Pragma split_var is specified on a variable whose type is not supported yet. "
+                        "Only unpacked 1D array is supported by this version.");
+    } else {  // finally find a good candidate
+        UINFO(4, m_lastVar->name() << " is added to candidate list.\n");
+        m_refs.insert(std::make_pair(m_lastVar, std::vector<AstArraySel*>()));
+    }
+    m_lastVar = NULL;
+    pragmap->unlinkFrBack()->deleteTree();  // consume the pragma here anyway.
+}
+
+void SplitVarVisitor::visit(AstArraySel* nodep) {
+    AstVarRef* const vrefp = VN_CAST(nodep->fromp(), VarRef);
+    if (!vrefp) return;
+
+    AstVar* const varp = vrefp->varp();
+    if (m_refs.find(varp) == m_refs.end()) return;  // variable without split_var pragma
+
+    AstConst* indexp = VN_CAST(nodep->bitp(), Const);
+    if (!indexp) {  // index is not constant yet.
+        UINFO(4, "Index of " << vrefp->name() << " is " << nodep->bitp()
+                             << " which is not constant.\n");
+        AstNode* const constified = V3Const::constifyEdit(nodep->bitp());
+        UINFO(4, "After constified:" << constified << '\n');
+        indexp = VN_CAST(constified, Const);
+    }
+
+    if (indexp) {  // OK
+        m_refs[varp].push_back(nodep);
+    } else {
+        nodep->bitp()->v3warn(SPLITVAR,
+                              "Variable \"" << vrefp->name() << "\" will not be splitted "
+                              "because index cannot be determined statically.");
+        m_refs.erase(varp);
+    }
+}
+
+// The actual splitting operation is done in this function, so don't forget to call this function.
+// The reason not doing things in dtor is to avoid throwing an exception in dtor.
+void SplitVarVisitor::split() {
+    for (vl_unordered_map<AstVar*, std::vector<AstArraySel*> >::iterator it = m_refs.begin(),
+                                                                         it_end = m_refs.end();
+        it != it_end; ++it) {
+        UINFO(3, "In module " << m_modp->name() << " var " << it->first->name() << " which has "
+                              << it->second.size() << " refs"
+                              << " will be splitted.\n");
+        AstVar* const varp = it->first;
+        AstUnpackArrayDType* const dtypep = VN_CAST(varp->subDTypep(), UnpackArrayDType);
+        std::vector<AstVar*> vars;
+        // Add the splitted variables
+        for (uint32_t i = 0; i < dtypep->arrayUnpackedElements(); ++i) {
+            std::stringstream name;
+            name << varp->name() << std::dec << i;
```

Comment:
string name = varp->name + "__BRA__" + cvtToStr(i) + "__KET__"

Also for loop needs fix to index based on subscripts of user, change test to be non-zero and signed e.g. [5:-2].

Reply:
I have updated the test to have such indices.
V3Width:width() removes VAR_BASE attribute and make index 0-origin so the original behavior was also good. I added a comment about it.